### ISSUE TITLE
API request loop without a memory overload

### ISSUE THREAD

**Issue by SNOWGUM** [axios/axios#2021]
This might seem strange but I'm trying to refresh an API as many times as possible until a `response.data` value reaches a certain number. Here's my code below<br>

    const axios = require('axios');
    const api = require('./api.json');
    const infinity = 0;
    
    let url = 'http://api.site.com/api.php?i=${api[item].appid}';
    
    function api_call(item){
      axios.get(url)
        .then(response => {
          let id = response.data.id;
          if (id > 10){
            return true;
          }else{
            return false;
          }
        })
        .catch(error => {
          console.log(error);
        });
      }
    
      function loop(variable) {
        while (infinity < 1) {
          if (api_call(variable) > true) {
            break;
            console.log('Found');
          }
          console.log('-');
        }
      }
This might seem strange, but obviously, I've modified it although this code would still achieve a similar goal to what I'm trying to accomplish with my code. Basically, it monitors an API until a value change triggers an action. Possibly I'm doing something wrong but when I run my version of this code (only changes are the API and value names) I start using massive amounts of Memory to the point that the program stops and gives me this error

    <--- Last few GCs --->
    
    [11524:000001E5724778C0]    93390 ms: Mark-sweep 1395.2 (1426.2) -> 1394.5 (1425.7) MB, 629.4 / 0.0 ms  (average mu = 0.139, current mu = 0.053) allocation failure scavenge might not succeed
    [11524:000001E5724778C0]    94039 ms: Mark-sweep 1395.4 (1425.7) -> 1394.8 (1426.2) MB, 638.3 / 0.0 ms  (average mu = 0.079, current mu = 0.016) allocation failure scavenge might not succeed
    
    
    <--- JS stacktrace --->
along with more information, I believe this is due to how I'm collecting the data, possibly there's some procedure I'm forgetting about I'm not so sure and can't find any similar topics regarding this. Thanks in advance for any help. If more information's required just ask I'll add more. 
[![enter image description here][1]][1]

  [1]: https://i.stack.imgur.com/MwuZW.png

**Comment by xgui3783** [comment]
Hmm, `api_call` will always return `undefined` You might want to use Promise or async

**Comment by SNOWGUM** [comment]
Sorry, I'm still not very familiar with the language. How would I incorporate this to work with a URL? If possible could you create an example with a fake API just so I can understand how that would work? If not that's ok I can just study it deeper :) @xgui3783

**Comment by xgui3783** [comment]
I will assume in your root directory, that there is a `json.json` file, and it looks like:

```
[{
  "name": "my-name-1",
  "id": "my-id-1"
},{
  "name":"my-name-2",
  "id":"my-id-2"
}]
```
and in the root directory, you also have an `index.js`:

```
const fs = require('fs')
const axios = require('axios')

const apiCall = async (url) => {
  const result = await axios.get(url)
  if (result.data === 'value I am looking for') {
    return result
  } else {
    return false
  }
}

const main = async () => {
  fs.readFile('./json.json', 'utf-8', (err, data) => {
    if (err) throw err
    const json = JSON.parse(data)
    let index = 0, result = null
    while (json[index]) {
      const url = 'http://api.site.com/api.php?id=' + json[index].id
      result = await apiCall(url)
      if (result) break;
      index += 1
    }
    console.log('result': result)
  })
}

main()

```

**Comment by SNOWGUM** [comment]
Well, no @xgui3783, I have JSON files for calling information like a config.json, but ultimately the data I want is from the web API. That's what my goal is, to get the data from the API as often as possible then once that data changes or is equal to a value run a separate command. Basically a while loop with an API. Although when I tried the code I posted I was using way too much memory. 
P.S, thanks for responding this issue is really irritating me :)
